File: c++/instance.cpp
#include "instance.h"

Job::Job(int index, vector<int>& task_sequence, int release_date, int due_date, float weight)
{
    _index = index;
    _task_sequence = task_sequence;
    _release_date = release_date;
    _due_date = due_date;
    _weight = weight;
}

Task::Task(int index, int processing_time, vector<int>& machines)
{
    _index = index;
    _processing_time = processing_time;
    _machines = machines;
}

Instance::Instance(int nb_operators, float alpha, float beta, vector<Job>& jobs, vector<Task>& tasks, vector<vector<vector<int>>>& operators)
{
    _nb_operators = nb_operators;
    _alpha = alpha;
    _beta = beta;
    _jobs = jobs;
    _tasks = tasks;
    _operators = operators;
}

std::ostream& operator<<(std::ostream& os, const Instance& instance) { 
    os << "Nb jobs: " << instance.nb_jobs() << endl;
    os << "Nb tasks: " << instance.nb_tasks() << endl;
    os << "Nb operators: " << instance.nb_operators() << endl;
    os << "Nb machines: " << instance.nb_machines();
    return os;
}

File: src/instance.jl
"""
    Job

## Attributes
- `index`: j
- `task_sequence`: list of task indices i
- `due_date`: dⱼ
- `weight`: wⱼ
"""
Base.@kwdef struct Job
    index::Int
    task_sequence::Vector{Int}
    release_date::Int
    due_date::Int
    weight::Float64
end

"""
    Task

## Attributes
- `index`: i
- `processing_time`: pᵢ
- `machines`: list of machine indices on which this task can be performed
"""
Base.@kwdef struct Task
    index::Int
    processing_time::Int
    machines::Vector{Int}
end

"""
    Instance

## Attributes
- `nb_operators::Int`
- `α::Float64`: penalty when a job is late
- `β::Float64`: delay penalty for jobs
- `jobs::Vector{Job}`: job list, in index order
- `tasks::Vector{Task}`: task list, in index order
- `operators::Matrix{Vector{Int}}`: operators[i, m] = list of operators that can operate task i on machine m
"""
Base.@kwdef struct Instance
    nb_operators::Int
    α::Float64 # unit penalty
    β::Float64 # tardiness
    jobs::Vector{Job}
    tasks::Vector{Task}
    operators::Matrix{Vector{Int}}  # operators[i, m] = list of operators that can operate task i on machine m
end

nb_jobs(instance::Instance) = length(instance.jobs)
nb_tasks(instance::Instance) = length(instance.tasks)
nb_machines(instance::Instance) = size(instance.operators, 2)
nb_operators(instance::Instance) = instance.nb_operators
